source/particles/particle_handler.cc

```diff
@@ -1485,6 +1549,127 @@ namespace Particles
+
+    std::vector<char> send_data;
+
+    // Fill data to send
+    if (send_pointers.back() > 0)
+      {
+        // Allocate space for sending particle data
+        send_data.resize(send_pointers.back());
+        void *data = static_cast<void *>(&send_data.front());
+
+        // Serialize the data sorted by receiving process
+        for (const auto i : neighbors)
+          for (const auto &p : particles_to_send.at(i))
+            {
+              p->write_data(data);
+              if (store_callback)
+                data = store_callback(p, data);
+            }
+      }
+
+    // Set up the space for the received particle data
+    std::vector<char> recv_data(recv_pointers.back());
+
+    // Exchange the particle data between domains
+    {
+      std::vector<MPI_Request> requests(2 * neighbors.size());
+      unsigned int             send_ops = 0;
+      unsigned int             recv_ops = 0;
+
+      const int mpi_tag = Utilities::MPI::internal::Tags::
+        particle_handler_send_recv_particles_send;
+
+      for (unsigned int i = 0; i < neighbors.size(); ++i)
+        if ((recv_pointers[i + 1] - recv_pointers[i]) > 0)
+          {
+            const int ierr =
+              MPI_Irecv(recv_data.data() + recv_pointers[i],
+                        recv_pointers[i + 1] - recv_pointers[i],
+                        MPI_CHAR,
+                        neighbors[i],
+                        mpi_tag,
+                        parallel_triangulation->get_communicator(),
+                        &(requests[send_ops]));
+            AssertThrowMPI(ierr);
+            send_ops++;
+          }
+
+      for (unsigned int i = 0; i < neighbors.size(); ++i)
+        if ((send_pointers[i + 1] - send_pointers[i]) > 0)
+          {
+            const int ierr =
+              MPI_Isend(send_data.data() + send_pointers[i],
+                        send_pointers[i + 1] - send_pointers[i],
+                        MPI_CHAR,
+                        neighbors[i],
+                        mpi_tag,
+                        parallel_triangulation->get_communicator(),
+                        &(requests[send_ops + recv_ops]));
+            AssertThrowMPI(ierr);
+            recv_ops++;
+          }
+      const int ierr =
+        MPI_Waitall(send_ops + recv_ops, requests.data(), MPI_STATUSES_IGNORE);
+      AssertThrowMPI(ierr);
+    }
+
+    // Put the received particles into the domain if they are in the
+    // triangulation
+    const void *recv_data_it = static_cast<const void *>(recv_data.data());
+
+    // Gather ghost particle iterators from the cache
+    auto &ghost_particles_iterators =
+      ghost_particles_cache.ghost_particles_iterators;
+    unsigned int particle_iterator_counter = 0;
+
+    while (reinterpret_cast<std::size_t>(recv_data_it) -
+             reinterpret_cast<std::size_t>(recv_data.data()) <
+           recv_pointers.back())
+      {
+        auto recv_particle =
```

Comment:
This is totally not your fault, but I wonder if these weird casts of pointers are the best approach. Maybe someone with more insight into the particles can explain me why this actually works?

Reply:
I must be honest and say I have no clue how these work... :(

Reply:
I think we should replace this weird structure and simply iterate over all entries in `ghost_particles_iterators`.

Reply:
If that can be made to work, I would very much prefer it.

Reply:
@gassmoeller  I would presume this originates from you? It would be nice to have your input here :)

Reply:
For the loop in `send_recv_particles_positions_and_properties` I can modify it to use the vector to loop instead. For the regular `send_recv_particles` I don't have an alternative though :).